chain/rust/src/builders/witness_builder.rs

```diff
@@ -0,0 +1,698 @@
+use std::{collections::{HashMap, BTreeSet}, fmt::Debug};
+use itertools::Itertools;
+use linked_hash_map::LinkedHashMap;
+
+use cml_crypto::{
+    DatumHash, Ed25519KeyHash, ScriptHash, RawBytesEncoding, Ed25519Signature, PublicKey
+};
+use crate::{
+    NativeScript, Script,
+    byron::ByronAddress,
+    crypto::{BootstrapWitness, Vkey, Vkeywitness, hash::hash_plutus_data},
+    transaction::{RequiredSigners, TransactionWitnessSet},
+    plutus::{PlutusData, PlutusScript, PlutusV1Script, PlutusV2Script, Redeemer, RedeemerTag, self},
+};
+
+use super::redeemer_builder::RedeemerWitnessKey;
+
+#[derive(Debug, thiserror::Error)]
+pub enum WitnessBuilderError {
+    #[error("Missing the following witnesses: {0:?}")]
+    MissingWitnesses(RequiredWitnessSet),
+}
+
+#[derive(Debug, Clone)]//, Eq, Ord, PartialEq, PartialOrd)]
+pub enum PlutusScriptWitness {
+    Ref(ScriptHash),
+    Script(PlutusScript),
+}
+
+impl PlutusScriptWitness {
+    // pub fn script(&self) -> Option<PlutusScript> {
+    //     match self {
+    //         Self::Ref(_) => None,
+    //         Self::Script(script) => Some(script.clone()),
+    //     }
+    // }
+
+    pub fn hash(&self) -> ScriptHash {
+        match self {
+            Self::Ref(hash) => hash.clone(),
+            Self::Script(script) => script.hash(),
+        }
+    }
+}
+
+impl From<PlutusScript> for PlutusScriptWitness {
+    fn from(script: PlutusScript) -> Self {
+        PlutusScriptWitness::Script(script)
+    }
+}
+
+impl From<ScriptHash> for PlutusScriptWitness {
+    fn from(hash: ScriptHash) -> Self {
+        PlutusScriptWitness::Ref(hash)
+    }
+}
+
+/// A partial Plutus witness
+/// It contains all the information needed to witness the Plutus script execution
+/// except for the redeemer tag and index
+/// Note: no datum is attached because only input script types have datums
+#[derive(Clone, Debug)]
+pub struct PartialPlutusWitness {
+    pub script: PlutusScriptWitness,
+    pub data: PlutusData,
+}
+
+#[derive(Clone, Debug)]
+pub enum InputAggregateWitnessData {
+    NativeScript(NativeScript, NativeScriptWitnessInfo),
+    PlutusScript(PartialPlutusWitness, RequiredSigners, Option<PlutusData>)
+}
+
+impl InputAggregateWitnessData {
+    pub fn plutus_data(&self) -> Option<&PlutusData> {
+        match self {
+            InputAggregateWitnessData::PlutusScript(witness, _, _) => {
+                Some(&witness.data)
+            }
+            _ => None
+        }
+    }
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct RequiredWitnessSet {
+    // note: the real key type for these is Vkey
+    // but cryptographically these should be equivalent and Ed25519KeyHash is more flexible
+    pub(crate) vkeys: BTreeSet<Ed25519KeyHash>,
```

Comment:
What do you think about allowing the access to the fields?

Reply:
Done as discussed.